unittest/frames-derivatives.cpp

```diff
@@ -0,0 +1,389 @@
+  Data::Matrix6x J_ref_local(6,model.nv); J_ref_local.setZero();
+  computeJointJacobians(model,data_ref,q);
+  getFrameJacobian(model,data_ref,frameId,WORLD,J_ref);
+  getFrameJacobian(model,data_ref,frameId,LOCAL_WORLD_ALIGNED,J_ref_local_world_aligned);
+  getFrameJacobian(model,data_ref,frameId,LOCAL,J_ref_local);
+  
+  BOOST_CHECK(data_ref.oMf[frameId].isApprox(data.oMf[frameId]));
+  BOOST_CHECK(partial_dv.isApprox(J_ref));
+  BOOST_CHECK(partial_dv_local_world_aligned.isApprox(J_ref_local_world_aligned));
+  BOOST_CHECK(partial_dv_local.isApprox(J_ref_local));
+  
+  // Check against finite differences
+  Data::Matrix6x partial_dq_fd(6,model.nv); partial_dq_fd.setZero();
+  Data::Matrix6x partial_dq_fd_local_world_aligned(6,model.nv); partial_dq_fd_local_world_aligned.setZero();
+  Data::Matrix6x partial_dq_fd_local(6,model.nv); partial_dq_fd_local.setZero();
+  Data::Matrix6x partial_dv_fd(6,model.nv); partial_dv_fd.setZero();
+  Data::Matrix6x partial_dv_fd_local_world_aligned(6,model.nv); partial_dv_fd_local_world_aligned.setZero();
+  Data::Matrix6x partial_dv_fd_local(6,model.nv); partial_dv_fd_local.setZero();
+  const double alpha = 1e-8;
+
+  // dvel/dv
+  Eigen::VectorXd v_plus(v);
+  Data data_plus(model);
+  forwardKinematics(model,data_ref,q,v);
+  Motion v0 = getFrameVelocity(model,data,frameId,WORLD);
+  Motion v0_local_world_aligned = getFrameVelocity(model,data,frameId,LOCAL_WORLD_ALIGNED);
+  Motion v0_local = getFrameVelocity(model,data,frameId,LOCAL);
+  for(int k = 0; k < model.nv; ++k)
+  {
+    v_plus[k] += alpha;
+    forwardKinematics(model,data_plus,q,v_plus);
+
+    partial_dv_fd.col(k) = (getFrameVelocity(model,data_plus,frameId,WORLD) - v0).toVector()/alpha;
+    partial_dv_fd_local_world_aligned.col(k) = (getFrameVelocity(model,data_plus,frameId,LOCAL_WORLD_ALIGNED) - v0_local_world_aligned).toVector()/alpha;
+    partial_dv_fd_local.col(k) = (getFrameVelocity(model,data_plus,frameId,LOCAL) - v0_local).toVector()/alpha;
+    v_plus[k] -= alpha;
+  }
+
+  BOOST_CHECK(partial_dv.isApprox(partial_dv_fd,sqrt(alpha)));
+  BOOST_CHECK(partial_dv_local_world_aligned.isApprox(partial_dv_fd_local_world_aligned,sqrt(alpha)));
+  BOOST_CHECK(partial_dv_local.isApprox(partial_dv_fd_local,sqrt(alpha)));
+
+  // dvel/dq
+  Eigen::VectorXd q_plus(q), v_eps(Eigen::VectorXd::Zero(model.nv));
+  forwardKinematics(model,data_ref,q,v);
+  updateFramePlacements(model,data_ref);
+
+  for(int k = 0; k < model.nv; ++k)
+  {
+    v_eps[k] += alpha;
+    q_plus = integrate(model,q,v_eps);
+    forwardKinematics(model,data_plus,q_plus,v);
+    updateFramePlacements(model,data_plus);
+
+    Motion v_plus_local_world_aligned = getFrameVelocity(model,data_plus,frameId,LOCAL_WORLD_ALIGNED);
+    SE3::Vector3 trans = data_plus.oMf[frameId].translation() - data_ref.oMf[frameId].translation();
+    v_plus_local_world_aligned.linear() -= v_plus_local_world_aligned.angular().cross(trans);
```

Comment:
It seems that my error was in the numdiff code. I didn't add this transformation.
Could you explain me why this is needed?

Reply:
The frame is translating. So you need to account for this change to express everything in the right coordinate system.